File: split/getnext.pas
procedure getnext;
label
  20, 25, 21, 26, 40, 10;
var
  k: 0..bufsize;
  t: halfword;
  cat: 0..15;
  c, cc: ASCIIcode;
  d: 2..3;
begin
  20:
    curcs := 0;
  if curinput.statefield <> 0 then{343:}

  begin
    25:
      if curinput.locfield <= curinput.limitfield then
      begin
        curchr := buffer[curinput.locfield];
        curinput.locfield := curinput.locfield + 1;
      21:
        curcmd := eqtb[3983 + curchr].hh.rh;
        {344:}
case curinput.statefield + curcmd of
          {345:}
10, 26, 42, 27, 43{:345}
: goto25;
          1, 17, 33:{354:}

          begin
            if curinput.locfield > curinput.limitfield then
              curcs := 513
            else
            begin
            26:
              k := curinput.locfield;
              curchr := buffer[k];
              cat := eqtb[3983 + curchr].hh.rh;
              k := k + 1;
              if cat = 11 then
                curinput.statefield := 17
              else if cat = 10 then
                curinput.
                statefield := 17
              else
                curinput.statefield := 1;
              if (cat = 11) and (k <= curinput.limitfield) then{356:}

              begin
                repeat
                  curchr :=
                    buffer[k];
                  cat := eqtb[3983 + curchr].hh.rh;
                  k := k + 1;
                until (cat <> 11) or (k > curinput.limitfield);
                {355:}
begin
                  if buffer[k] = curchr then
                    if cat = 7 then
                      if k < curinput.limitfield then
                      begin
                        c := buffer[k + 1];
                        if c < 128 then
                        begin
                          d := 2;
                          if (((c >= 48) and (c <= 57)) or ((c >= 97) and (c <= 102))) then
                            if k + 2 <= curinput.limitfield then
                            begin
                              cc := buffer[k + 2];
                              if (((cc >= 48) and (cc <= 57)) or ((cc >= 97) and (cc <= 102))) then
                                d := d + 1;
                            end;
                          if d > 2 then
                          begin
                            if c <= 57 then
                              curchr := c - 48
                            else
                              curchr := c - 87;
                            if cc <= 57 then
                              curchr := 16 * curchr + cc - 48
                            else
                              curchr := 16 * curchr + cc - 87;
                            buffer[k - 1] := curchr;
                          end
                          else if c < 64 then
                            buffer[k - 1] := c + 64
                          else
                            buffer[k - 1] := c - 64;
                          curinput.limitfield := curinput.limitfield - d;
                          First := First - d;
                          while k <= curinput.limitfield do
                          begin
                            buffer[k] := buffer[k + d];
                            k := k + 1;
                          end;
                          goto 26;
                        end;
                      end;
                end{:355}
;
                if cat <> 11 then
                  k := k - 1;
                if k > curinput.locfield + 1 then
                begin
                  curcs := idlookup(curinput.locfield, k - curinput.locfield);
                  curinput.locfield := k;
                  goto 40;
                end;
              end{:356}

              else{355:}
if buffer[k] = curchr then
                if cat = 7 then
                  if k < curinput.limitfield then
                  begin
                    c := buffer[k + 1];
                    if c < 128 then
                    begin
                      d := 2;
                      if (((c >= 48) and (c <= 57)) or ((c >= 97) and (c <= 102))) then
                        if k + 2 <= curinput.limitfield then
                        begin
                          cc := buffer[k + 2];
                          if (((cc >= 48) and (cc <= 57)) or ((cc >= 97) and (cc <= 102))) then
                            d := d + 1;
                        end;
                      if d > 2 then
                      begin
                        if c <= 57 then
                          curchr := c - 48
                        else
                          curchr := c - 87;
                        if cc <= 57 then
                          curchr := 16 * curchr + cc - 48
                        else
                          curchr := 16 * curchr + cc - 87;
                        buffer[k - 1] := curchr;
                      end
                      else if c < 64 then
                        buffer[k - 1] := c + 64
                      else
                        buffer[k - 1] := c - 64;
                      curinput.limitfield := curinput.limitfield - d;
                      First := First - d;
                      while k <= curinput.limitfield do
                      begin
                        buffer[k] := buffer[k + d];
                        k := k + 1;
                      end;
                      goto 26;
                    end;
                  end{:355}
;
              curcs := 257 + buffer[curinput.locfield];
              curinput.locfield := curinput.locfield + 1;
            end;
          40:
            curcmd := eqtb[curcs].hh.b0;
            curchr := eqtb[curcs].hh.rh;
            if curcmd >= 113 then
              checkoutervalidity;
          end{:354}
;
          14, 30, 46:{353:}

          begin
            curcs := curchr + 1;
            curcmd := eqtb[curcs].hh.b0;
            curchr := eqtb[curcs].hh.rh;
            curinput.statefield := 1;
            if curcmd >= 113 then
              checkoutervalidity;
          end{:353}
;
          8, 24, 40:{352:}

          begin
            if curchr = buffer[curinput.locfield] then
              if curinput.locfield < curinput.limitfield then
              begin
                c := buffer[curinput.locfield + 1];
                if c < 128 then
                begin
                  curinput.locfield := curinput.locfield + 2;
                  if (((c >= 48) and (c <= 57)) or ((c >= 97) and (c <= 102))) then
                    if curinput.locfield <= curinput.limitfield then
                    begin
                      cc := buffer[curinput.locfield];
                      if (((cc >= 48) and (cc <= 57)) or ((cc >= 97) and (cc <= 102))) then
                      begin
                        curinput.
                        locfield := curinput.locfield + 1;
                        if c <= 57 then
                          curchr := c - 48
                        else
                          curchr := c - 87;
                        if cc <= 57 then
                          curchr := 16 * curchr + cc - 48
                        else
                          curchr := 16 * curchr + cc - 87;
                        goto 21;
                      end;
                    end;
                  if c < 64 then
                    curchr := c + 64
                  else
                    curchr := c - 64;
                  goto 21;
                end;
              end;
            curinput.statefield := 1;
          end{:352}
;
          16, 32, 48:{346:}

          begin
            begin
              if interaction = 3 then;
              printnl(262);
              print(613);
            end;
            begin
              helpptr := 2;
              helpline[1] := 614;
              helpline[0] := 615;
            end;
            deletionsallowed := false;
            error;
            deletionsallowed := true;
            goto 20;
          end{:346}
;
          {347:}
11:{349:}

          begin
            curinput.statefield := 17;
            curchr := 32;
          end{:349}
;
          6:{348:}

          begin
            curinput.locfield := curinput.limitfield + 1;
            curcmd := 10;
            curchr := 32;
          end{:348}
;
          22, 15, 31, 47:{350:}

          begin
            curinput.locfield := curinput.limitfield + 1;
            goto 25;
          end{:350}
;
          38:{351:}

          begin
            curinput.locfield := curinput.limitfield + 1;
            curcs := parloc;
            curcmd := eqtb[curcs].hh.b0;
            curchr := eqtb[curcs].hh.rh;
            if curcmd >= 113 then
              checkoutervalidity;
          end{:351}
;
          2: alignstate := alignstate + 1;
          18, 34:
          begin
            curinput.statefield := 1;
            alignstate := alignstate + 1;
          end;
          3: alignstate := alignstate - 1;
          19, 35:
          begin
            curinput.statefield := 1;
            alignstate := alignstate - 1;
          end;
          20, 21, 23, 25, 28, 29, 36, 37, 39, 41, 44, 45: curinput.statefield := 1;
          {:347}
others: end{:344}
;
      end
      else
      begin
        curinput.statefield := 33;
        {360:}
if curinput.namefield > 17 then{362:}

        begin
          line := line + 1;
          First := curinput.startfield;
          if not forceeof then
            if inputln(inputfile[curinput.indexfield], true) then
              firmuptheline
            else
              forceeof := true;
          if forceeof then
          begin
            printchar(41);
            openparens := openparens - 1;
            break(termout);
            forceeof := false;
            endfilereading;
            checkoutervalidity;
            goto 20;
          end;
          if (eqtb[5311].int < 0) or (eqtb[5311].int > 255) then
            curinput.limitfield :=
              curinput.limitfield - 1
          else
            buffer[curinput.limitfield] := eqtb[5311].int;
          First := curinput.limitfield + 1;
          curinput.locfield := curinput.startfield;
        end{:362}

        else
        begin
          if not (curinput.namefield = 0) then
          begin
            curcmd := 0;
            curchr := 0;
            goto 10;
          end;
          if inputptr > 0 then
          begin
            endfilereading;
            goto 20;
          end;
          if selector < 18 then
            openlogfile;
          if interaction > 1 then
          begin
            if (eqtb[5311].int < 0) or (eqtb[5311].int > 255) then
              curinput.limitfield := curinput.limitfield + 1;
            if curinput.limitfield = curinput.startfield then
              printnl(616);
            println;
            First := curinput.startfield;
            begin
              ;
              print(42);
              terminput;
            end;
            curinput.limitfield := last;
            if (eqtb[5311].int < 0) or (eqtb[5311].int > 255) then
              curinput.limitfield :=
                curinput.limitfield - 1
            else
              buffer[curinput.limitfield] := eqtb[5311].int;
            First := curinput.limitfield + 1;
            curinput.locfield := curinput.startfield;
          end
          else
            fatalerror(617);
        end{:360}
;
        begin
          if interrupt <> 0 then
            pauseforinstructions;
        end;
        goto 25;
      end;
  end{:343}

  else{357:}
if curinput.locfield <> 0 then
  begin
    t := mem[curinput.locfield].hh.lh;
    curinput.locfield := mem[curinput.locfield].hh.rh;
    if t >= 4095 then
    begin
      curcs := t - 4095;
      curcmd := eqtb[curcs].hh.b0;
      curchr := eqtb[curcs].hh.rh;
      if curcmd >= 113 then
        if curcmd = 116 then{358:}

        begin
          curcs := mem[curinput.locfield].hh.lh - 4095;
          curinput.locfield := 0;
          curcmd := eqtb[curcs].hh.b0;
          curchr := eqtb[curcs].hh.rh;
          if curcmd > 100 then
          begin
            curcmd := 0;
            curchr := 257;
          end;
        end{:358}

        else
          checkoutervalidity;
    end
    else
    begin
      curcmd := t div 256;
      curchr := t mod 256;
      case curcmd of
        1: alignstate := alignstate + 1;
        2: alignstate := alignstate - 1;
        5:{359:}

        begin
          begintokenlist(paramstack[curinput.limitfield + curchr - 1], 0);
          goto 20;
        end{:359}
;
        others: end;
    end;
  end
  else
  begin
    endtokenlist;
    goto 20;
  end{:357}
;
  {342:}
if curcmd <= 5 then
    if curcmd >= 4 then
      if alignstate = 0 then{789:}

      begin
        if (scannerstatus = 4) or (curalign = 0) then
          fatalerror(595);
        curcmd := mem[curalign + 5].hh.lh;
        mem[curalign + 5].hh.lh := curchr;
        if curcmd = 63 then
          begintokenlist(29990, 2)
        else
          begintokenlist(mem[curalign + 2].int, 2);
        alignstate := 1000000;
        goto 20;
      end{:789}
{:342}
;
  10: ;
end;

File: cpp/src/getnext.cpp
#include "getnext.h"
#include "impression.h"
#include "erreur.h"
#include "fichier.h"
#include "lecture.h"
#include "chaine.h"
#include "noeud.h"
#include "police.h"
#include "equivalent.h" 
#include "alignement.h"
#include "macrocall.h"
#include "conditional.h"
#include "etat.h"
#include "initprim.h"
#include "tampon.h"
#include <iostream>

static Token checkoutervalidity(char status, Token t)
{
	if (status == normal)
		return t;
	if (t.cs && (state == token_list || !isFile()))
		backList(cs_token_flag+t.cs);
	switch (status)
	{
		case skipping:
			inserror(t, "Incomplete "+Token(if_test, curif).cmdchr()+"; all text was ignored after line "+std::to_string(skipline), std::string(cs ? "A forbidden control sequence occurred in skipped text." : "The file ended while I was skipping conditional text.")+"This kind of error happens when you say `\\if...' and forget\nthe matching `\\fi'. I've inserted a `\\fi'; this might work.", false);
			return frozen_fi+cs_token_flag;
		case defining:
			print("\rRunaway definition?\n"+defRef.toString(errorline-10));
			error(t.cs ? "Forbidden control sequence found" : "File ended while scanning definition of "+scs(warningindex), "I suspect you have forgotten a `}', causing me\nto read past where you wanted me to stop.\nI'll try to recover; but if the error is serious,\nyou'd better type `E' or `X' now and fix your file.", false);
			insList(right_brace_token+'}');
			break;
		case matching:
			print("\rRunaway argument?\n"+tempHead.toString(errorline-10));
			error(t.cs ? "Forbidden control sequence found" : "File ended while scanning use of "+scs(warningindex), "I suspect you have forgotten a `}', causing me\nto read past where you wanted me to stop.\nI'll try to recover; but if the error is serious,\nyou'd better type `E' or `X' now and fix your file.", false);
			longstate = outer_call;
			insList(partoken);
			break;
		case aligning:
			print("\rRunaway preamble?\n"+holdHead.toString(errorline-10));
			error(t.cs ? "Forbidden control sequence found" : "File ended while scanning preamble of "+scs(warningindex), "I suspect you have forgotten a `}', causing me\nto read past where you wanted me to stop.\nI'll try to recover; but if the error is serious,\nyou'd better type `E' or `X' now and fix your file.", false);
			insList(std::vector<halfword>{frozen_cr+cs_token_flag, right_brace_token+'}'});
			alignstate = -1000000;
			break;
		case absorbing:
			print("\rRunaway text?\n"+defRef.toString(errorline-10));
			error(t.cs ? "Forbidden control sequence found" : "File ended while scanning text of "+scs(warningindex), "I suspect you have forgotten a `}', causing me\nto read past where you wanted me to stop.\nI'll try to recover; but if the error is serious,\nyou'd better type `E' or `X' now and fix your file.", false);
			insList(right_brace_token+'}');
	}
	return t.cs ? Token(spacer, ' ') : t;
}

static bool is_hex(ASCIIcode c) { return between('0', c, '9') || between('a', c, 'f'); }
static int toHex(ASCIIcode c) { return c <= '9' ? c -'0' : c-'a'+10; }

static int intFromTwoHexDigit(ASCIIcode c, int pos)
{
	if (is_hex(c) && pos <= curinput.limit)
	{
		auto &cc = getBuf(pos);
		if (is_hex(cc))
			return (toHex(c)<<4)+toHex(cc);
	}
	return -1;
}

static bool prochainCaractereOK(int pos, halfword chr) { return getBuf(pos) == chr && pos < curinput.limit && getBuf(pos+1) < 128; }

template<class T> static int processBuffer(int k, T &var)
{
	auto &c = getBuf(k+1);
	auto t = intFromTwoHexDigit(c, k+2);
	var = t >= 0 ? t : c < 64 ? c + 64 : c - 64;
	return t >= 0 ? 3 : 2;
}

static void removeFromEnd(int &k, int d)
{
	curinput.limit -= d;
	First -= d;
	shiftBuf(k+d, curinput.limit+1+d, k);
	k = curinput.limit+1;
}

static bool testDoubleSup(int k, Token &tk)
{
	if (k >= curinput.limit)
		return false;
	if (getBuf(k) == tk.chr /* ^^ */ && getBuf(k+1) < 128)
	{
		auto &c = getBuf(k+1);
		auto t = intFromTwoHexDigit(c, k+2);
		tk.chr = getBuf(k-1) = t >= 0 ? t : c < 64 ? c + 64 : c - 64;
		if (t >= 0) // ^^xx avec xx hexadécimal
			removeFromEnd(k, 3);
		else // ^^c avec c caractère
			removeFromEnd(k, 2);
		return true;
	}
	return false;
}


static TokenList omit_template(end_template_token); //!< a constant token list

#define ANY_STATE_PLUS(cmd) mid_line+cmd: case skip_blanks+cmd: case new_line+cmd
#define ADD_DELIMS_TO(state) state+math_shift: case state+tab_mark: case state+mac_param: case state+sub_mark: case state+letter: case state+other_char 

Token Scanner::next(char status, bool nonewcontrolsequence)
{
	while (true)
	{
		bool restart = false;
		Token t;
		if (state != token_list)
		{
			bool skip;
			do
			{
				skip = false;
				if (!isBufFull())
				{
					t.chr = currentBuf();
					bufNext();
					bool doubleSup2;
					do
					{
						doubleSup2 = false;
						t.cmd = cat_code(t.chr);
						switch (state+t.cmd)
						{
							case ANY_STATE_PLUS(ignore):
							case skip_blanks+spacer:
							case new_line+spacer:
								skip = true;
								break;
							case ANY_STATE_PLUS(escape):
								if (isBufFull())
									t = equivToken(null_cs-active_base, eqtb_active);
								else
								{   // loc <= limit
									bool identifiant = false;
									for (bool doubleSup = true; doubleSup && !identifiant;)
									{
										doubleSup = false;
										switch (int cat = cat_code(t.chr = currentBuf()); cat)
										{
											case letter:
												if (atLeastTwoInBuf())
												{
													int k;
													for (k = curinput.loc+1; cat == letter && k <= curinput.limit; k++)
														cat = cat_code(t.chr = getBuf(k));
													if (cat == sup_mark && testDoubleSup(k, t)) // "^^" 
														doubleSup = true;
													else
													{
														if (cat != letter)
															k--;
														if (k-curinput.loc >= 1)
														{
															t = equivToken(idlookup(readBufUpTo(k), nonewcontrolsequence)-hash_base, eqtb_cs);
															curinput.loc = k;
															identifiant = true;
														}
													}
												} [[fallthrough]];
											case spacer:
												state = skip_blanks;
												break;
											case sup_mark:
												if (testDoubleSup(curinput.loc+1, t))
													doubleSup = true; [[fallthrough]];
											default:
												state = mid_line;
										}
									}
									if (!identifiant)
									{
										t = equivToken(currentBuf()+single_base-active_base, eqtb_active);
										bufNext();
									}
								}
								if (t.cmd >= outer_call)
									t = checkoutervalidity(status, t);
								break;
							case ANY_STATE_PLUS(active_char):
								t = equivToken(t.chr, eqtb_active);
								state = mid_line;
								if (t.cmd >= outer_call)
									t = checkoutervalidity(status, t);
								break;
							case ANY_STATE_PLUS(sup_mark):
								if (prochainCaractereOK(curinput.loc, t.chr)) 
								{
									curinput.loc += processBuffer(curinput.loc, t.chr);
									doubleSup2 = true;
								}
								else
									state = mid_line;
								break;
							case ANY_STATE_PLUS(invalid_char):
								error("Text line contains an invalid character", "A funny symbol that I can't read has just been input.\nContinue, and I'll forget that it ever happened.", false);
								restart = true;
								break;
							case mid_line+spacer:
								state = skip_blanks;
								t.chr = ' ';
								break;
							case mid_line+car_ret:
								bufEnd();
								t.cmd = spacer;
								t.chr = ' ';
								break;
							case skip_blanks+car_ret:
							case ANY_STATE_PLUS(comment):
								bufEnd();
								skip = true;
								break;
							case new_line+car_ret:
								bufEnd();
								t = equivToken(parloc, eqtb);
								if (t.cmd >= outer_call)
									t = checkoutervalidity(status, t);
								break;
							case skip_blanks+left_brace:
							case new_line+left_brace:
								state = mid_line;
							case mid_line+left_brace:
								alignstate++;
								break;
							case skip_blanks+right_brace:
							case new_line+right_brace:
								state = mid_line;
							case mid_line+right_brace:
								alignstate--;
								break;
							case ADD_DELIMS_TO(skip_blanks):
							case ADD_DELIMS_TO(new_line):
								state = mid_line;
						}
					} while (doubleSup2 && !skip && !restart);
				}
				else // loc > limit
				{
					state = new_line;
					if (isFile())
					{
						if (!terminal_input(name))
							return Token(0, 0);
						if (inputstack.size() > 1)
						{
							First = curinput.start;
							endfilereading();
							restart = true;
						}
						if (!restart)
						{
							if (selector < log_only)
								openlogfile(First);
							if (interaction > nonstop_mode)
							{
								if (end_line_char_inactive())
									curinput.limit++;
								if (sizeBuf() == 1)
									print("\r(Please type a command or say `\\end')");
								println();
								print("*");
								curinput.limit = terminput(curinput.start);
								removeLastBuf();
								First = curinput.limit+1;
								bufBegin();
							}
							else
								fatalerror("*** (job aborted, no legal \end found)");
						}
					}
					else
					{
						line++;
						First = curinput.start;
						if (int lst = inputln(cur_file(), true, curinput.start); !forceeof && lst >= 0)
						{
							curinput.limit = lst;
							if (pausing() > 0 && interaction > nonstop_mode)
							{
								lst = firmuptheline();
								First = curinput.limit;
							}
							if (!restart)
							{
								removeLastBuf();
								First = curinput.limit+1;
								bufBegin();
							}
						}
						else
						{
							print(")");
							openparens--;
							std::cout << std::flush;
							forceeof = false;
							First = curinput.start;
							endfilereading();
							t = checkoutervalidity(status, t);
							restart = true;
						}
					}
					if (!restart)
						skip = true;
				}
			} while (skip && !restart);
		}
		else
			if (!Loc)
			{
				endtokenlist();
				restart = true;
			}
			else
			{
				Loc++;
				if (Start.list[Loc] >= cs_token_flag)
				{
					t = equivToken(Start.list[Loc]-cs_token_flag, eqtb);
					switch (t.cmd)
					{
						case outer_call:
						case long_outer_call:
						case end_template:
						case glue_ref:
						case shape_ref:
						case box_ref:
						case data:
							t = checkoutervalidity(status, t);
							break;
						case dont_expand:
						{
							t = equivToken(Start.list[Loc]-cs_token_flag, eqtb);
							Loc = 0;
							if (t.cmd > max_command)
								t = Token(relax, no_expand_flag);
						}
					}
				}
				else
					switch (t = Token(Start.list[Loc]>>8, Start.list[Loc]%(1<<8)); t.cmd)
					{
						case left_brace: 
							alignstate++;
							break;
						case right_brace: 
							alignstate--;
							break;
						case out_param:
							paramstack[curinput.limit+t.chr-1].beginBelowMacro(parameter);
							restart = true;
					}
			}
		if (!restart)
		{ 
			if (between(tab_mark, t.cmd, out_param) && alignstate == 0)
			{
				if (status == aligning || curalign == nullptr)
					fatalerror("(interwoven alignment preambles are not allowed)");
				t.cmd = curalign->extra_info;
				curalign->extra_info = t.chr;
				t.cmd == omit ? omit_template.beginBelowMacro(v_template) : curalign->vPart.beginBelowMacro(v_template);
				alignstate = 1000000;
				restart = true;
			}
			return t;
		}
	}
}
